### DISCUSSION TITLE
4. 데이터베이스 - 데이터 모델링, 정규화

### DISCUSSION THREAD

**Discussion by zini9188** [Shortudy-10th/tech-for-developer#153] [질의 응답]
- 데이터 모델링의 개념
- ER 모델
- ER 모델을 관계 데이터 모델로 사상
- 이상현상
- 함수 종속성
- 정규화
- 그 외 등등

**Comment by kjc6735** [comment]
데이터베이스 뷰의 장점과 단점을 설명해주세요

**Reply by kjc6735** [reply]
장잠으로는 논리적 데이터 독립성을 제공합니다. 또한 복잡한 쿼리를 단순화해서 사용이 가능하고, 데이터 관리와 보안에 용이합니다. 
단점은 삽입,삭제,갱신 연산에 많은 제약 따르고 인덱스를 구성할 수 없습니다. 정의된 뷰는 수정이 불가능합니다.

**Comment by Jihn0118** [comment]
## Q. 역정규화를 하는 이유에 대해 아는 대로 설명해주세요.

데이터베이스의 성능 향상과 쿼리 처리의 효율성을 위해 데이터 중복을 허용하고,
 데이터를 다시 구조화하는 역정규화를 하게 됩니다.

**Comment by zini9188** [comment]
### Q. 정규화에 대해 설명해주세요.

속성간의 종속성으로 인한 이상현상이 발생하는 릴레이션을 분해하여 재디자인함으로써 이상현상을 없애는 과정으로, 데이터의 중복 방지 및 무결성을 충족하기 위해 데이터베이스를 설계하는 방법입니다.

**Reply by seongeun42** [reply]
## Q. 이상현상이란 무엇인가요?

이상현상이란 잘못된 설계로 인해 데이터의 삽입, 삭제, 수정할 때 생기는 문제를 말합니다. **삽입 이상**은 어떤 데이터를 삽입하려고 할 때 불필요하고 원치 않는 데이터까지 함께 삽입해야 하는 것입니다. **삭제 이상**은 한 튜플을 삭제함으로써 유지되어야 할 데이터까지 삭제되는 연쇄 삭제 현상으로 정보 손실이 발생하는 것입니다. **수정 이상**은 중복된 튜플 중 일부 튜플의 값만 수정함으로써 무결성이 깨져 정보의 모순이 생기는 것입니다.

**Reply by seongeun42** [reply]
## Q. 정규화의 종류에 대해 말해주세요.

제 1 정규형은 모든 애트리뷰트가 원자값으로 이루어진 릴레이션입니다.
제 2 정규형은 제 1 정규형을 만족하고, 부분 함수 종속을 제거한 릴레이션입니다.
제 3 정규형은 제 2 정규형을 만족하고, 이행적 함수 종속을 제거한 릴레이션입니다.
보이스/코드 정규형은 제 3 정규형을 만족하고, 결정자가 후보키가 아닌 함수 종속을 제거한 릴레이션입니다.
제 4 정규형은 BCNF를 만족하고, 다치 종속을 제거한 릴레이션입니다.
제 5 정규형은 제 4 정규형을 만족하고, 후보키를 통하지 않은 조인 종속을 제거한 릴레이션입니다.

<img src="https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/36d005bb-314b-40aa-9206-427781261428" width=500>

**Reply by seongeun42** [reply]
## Q. 함수 종속성이란 무엇인가요?

A라는 값을 알면 B의 값도 알 수 있다고 한다면, B는 A에 종속되어 있다고 하며 A → B  라고 표현합니다. 이때, A는 결정자, B는 종속자라고 합니다.

함수 종속에는 하나의 값이 다른 하나의 값에 종속되어 있는 완전 함수 종속, 기본 키가 2개 이상의 복합 키로 되어 있을 때 복합 키 중 일부의 값에만 종속되어 있는 부분 함수 종속, A → C가 존재할 때, A → B와 B → C가 성립하는 이행 함수 종속이 있습니다.

**Comment by watty-exclamationMark** [comment]
## Q. 역정규화란 무엇인가요?
데이터베이스의 비용을 최소화하기 위해 중복을 허용하며 Entity를 다시 통합하거나 분할하여 정규화 과정을 통해 도출된 DB 구조를 재조정하는 과정입니다.

**Reply by watty-exclamationMark** [reply]
## 역정규화가 필요한 이유는 무엇인가요?

정규화를 거치면 JOIN 연산이 많아지는데, 이로 인해 요청을 처리하는 시간이 길어질 수 있습니다.
따라서 DB의 전반적인 성능을 향상시키기 위해 역정규화 과정이 필요할 수 있습니다.

**Comment by seongeun42** [comment]
## Q. 데이터 모델이란 무엇인가요?

데이터 모델은 논리적으로 표현된 데이터 구조, 이 구조에서 허용될 수 있는 연산, 이 구조와 연산에서의 제약조건에 대한 명세를 기술한 것입니다.

+) 데이터 모델의 3 요소 : 구조, 연산, 제약조건
